src/sentry/static/sentry/app/utils/apm.jsx

```diff
@@ -1,141 +1,33 @@
- *
- * TODO(apm): We probably want a hard limit for root span, e.g. it's possible we have long
- * API requests combined with renders that could create a very long root span.
- *
- * TODO(apm): Handle polling requests?
- */
-function interruptFlush() {
-  if (!flushTransactionTimeout) {
-    return;
-  }
-
-  clearTimeout(flushTransactionTimeout);
-  wasInterrupted = true;
-}
-
-export function finishTransaction(delay = TRANSACTION_TIMEOUT) {
-  if (flushTransactionTimeout || (hasActiveRenders() || hasActiveRequests())) {
-    interruptFlush();
-  }
-
-  flushTransactionTimeout = setTimeout(() => {
-    if (currentTransactionSpan) {
-      currentTransactionSpan.finish();
-      currentTransactionSpan = null;
-      firstPageLoad = false;
-    }
-  }, delay);
-}
-
-/**
- * These `start-` functions attempt to track the state of "actions".
- *
- * They interrupt the transaction flush (which times out), and
- * requires the related `finish-` function to be called.
- */
-export function startRequest(id) {
-  requests.add(id);
-  interruptFlush();
-}
-export function startRender(id) {
-  renders.add(id);
-  interruptFlush();
-}
-
-/**
- * These `finish-` functions clean up the "active" state of an ongoing "action".
- * If there are no other "actions" and we have interrupted a flush, we should
- * finish the transaction
- */
-export function finishRequest(id) {
-  requests.delete(id);
-  // TODO(apm): Is this necessary? flush should be interrupted already from start()
-  interruptFlush();
-
-  if (wasInterrupted && !hasActiveRenders() && !hasActiveRequests()) {
-    finishTransaction(1);
-  }
-}
-export function finishRender(id) {
-  renders.delete(id);
-  interruptFlush();
-
-  if (wasInterrupted && !hasActiveRenders() && !hasActiveRequests()) {
-    finishTransaction(1);
-  }
-}
+import {TransactionActivity} from '@sentry/integrations';
 
 /**
  * Sets the transaction name
  */
 export function setTransactionName(name) {
+  TransactionActivity.updateTransactionName(`${name}`);
   Sentry.configureScope(scope => {
-    const span = scope.getSpan();
-
-    if (!span) {
-      return;
-    }
-
-    span.transaction = firstPageLoad ? `PageLoad: ${name}` : name;
     scope.setTag('ui.route', name);
   });
 }
 
 /**
- * This is called only when our application is initialized. Creates a root span
- * and creates a router listener to create a new root span as user navigates.
+ * This is called only when our application is initialized. Creates a transaction
+ * and creates a router listener to create a new transaction as user navigates.
  */
 export function startApm() {
+  TransactionActivity.startIdleTransaction(`${window.location.href}`, {
+    op: 'pageload',
+    sampled: true,
+  });
   Sentry.configureScope(scope => {
-    currentTransactionSpan = Sentry.startSpan({
-      op: 'pageload',
+    scope.setTag('ui.nav', 'pageload');
+  });
+  Router.browserHistory.listen(() => {
+    TransactionActivity.startIdleTransaction(`${window.location.href}`, {
```

Comment:
hmm would be nice if we had the generic router path here

Reply:
Yeah, we don't at this point since the route is "lazy" loaded when the Router is mount here: https://github.com/getsentry/sentry/blob/master/src/sentry/static/sentry/app/views/app.jsx#L182

Reply:
Will we end up with many different transaction names for the same route with this approach?

Reply:
`setTransactionName` basically "overwrites" this right? can we add a comment about it?